trulens/nn/models/keras_utils.py

```diff
@@ -0,0 +1,354 @@
+    disconnected inbound nodes. To specify which inbound nodes belong to the 
+    provided model, this method returns a mapping of layers to the index of their 
+    respective inbound node.
+     
+    Parameters
+    ----------
+    model : keras.models.Model
+        Base model
+    
+    Returns
+    -------
+    Dict[str, int]
+    Returns a mapping of layer names to the index of the inbound node associated with model
+
+    '''
+    innode_indices = {}
+    nodes_by_depth = model._nodes_by_depth
+
+    def tracer(depth):
+        if depth not in nodes_by_depth:
+            return
+        nodes = nodes_by_depth[depth]
+        for node in nodes:
+            out_layer = node.outbound_layer
+            if hasattr(out_layer, "layers"):
+                # Is nested model, recurse
+                sub_innode_indices = trace_input_indices(out_layer)
+                for layer_name, idx in sub_innode_indices.items():
+                    innode_indices[f"{out_layer.name}/{layer_name}"] = idx
+
+            if node in out_layer.inbound_nodes:
+                idx = out_layer.inbound_nodes.index(node)
+                if out_layer.name in innode_indices and innode_indices[
+                        out_layer.name] != idx:
+                    # May occur if layer is shared between other layers in the same model
+                    print(
+                        f"{out_layer.name}: input node conflict. orig: {innode_indices[out_layer.name]}, new: {idx}. Keeping original"
+                    )
+                if out_layer.name not in innode_indices:
+                    innode_indices[out_layer.name] = idx
+        tracer(depth + 1)
+
+    tracer(0)
+    return innode_indices
+
+
+def load_keras_model_from_handle(
+    handle, orig_layer, keras_module, tfhub_module
+):
+    '''
+    load_keras_model_from_handle Load the tensorflow hub KerasLayer as a keras Model.
+     
+    Parameters
+    ----------
+    handle : str
+        TFHub handle
+
+    orig_layer : KerasLayer
+    
+    keras_module : Python Module
+        Either the keras module or tf.keras module
+    
+    tfhub_module : Python Module
+        tensorflow_hub module. Used to resolve tfhub model paths
+    
+    Returns
+    -------
+    keras.models.Model
+
+    '''
+    module_path = tfhub_module.module_v2.resolve(handle)
+    keras_model = keras_module.models.load_model(module_path)
+    keras_model.set_weights(orig_layer.get_weights())
+    keras_model._name = orig_layer.name
+
+    # Unnest models from input layers
+    while True:
+        if not hasattr(keras_model, "layers"):
+            break
+
+        non_input_layers = []
+        for layer in keras_model.layers:
+            if not isinstance(layer, keras_module.layers.InputLayer):
+                layer._name = f"{keras_model.name}/{layer.name}"
+                non_input_layers.append(layer)
+
+        if len(non_input_layers) != 1:
+            break
```

Comment:
If there's more than one non-input layer, it's possible they're not run sequentially. The logic about their run order and which inputs/output tensors are fed to the next layer are contained in the parent model, so I can't dive deeper.

This loop is there to strip away input layers from deeply nested models, since the inputs don't do much more than forward their values to the next depth's input layers.

Going from `[input *3 > [input * 3 > [encoder_layers]]]`
to just `[encoder_layers]`

Reply:
The model can be correctly defined with connected input and output tensors. However, Keras doesn't support taking gradients of layers in outer models wrt layers in nested models. Therefore, we need to flatten the architecture so that layers of interest are exposed in the outermost model.

I made a change to properly flatten nested models so that this loop isn't necessary. It uses the inner model's `.call()` method to apply the "inner model" logic onto the "outer model" context and expose nested layers. Hopefully should be cleaner/less hacky than stripping non-input layers and calling them individually. Let me know what you think!

Reply:
perfect, Thanks Corey!